crates/pallet-native-to-evm-currency-swap/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+    <<T as Config>::NativeCurrency as Inspect<<T as frame_system::Config>::AccountId>>::Balance;
+
+/// Utility alias for easy access to the [`Inspect::Balance`] of the [`Config::EvmCurrency`] type.
+type EvmBalanceOf<T> =
+    <<T as Config>::EvmCurrency as Inspect<<T as Config>::EvmAccountId>>::Balance;
+
+// We have to temporarily allow some clippy lints. Later on we'll send patches to substrate to
+// fix them at their end.
+#[allow(clippy::missing_docs_in_private_items)]
+#[frame_support::pallet]
+pub mod pallet {
+    use fp_ethereum::ValidatedTransaction;
+    use frame_support::{
+        pallet_prelude::*,
+        sp_runtime::traits::{Convert, MaybeDisplay, UniqueSaturatedInto},
+        sp_std::fmt::Debug,
+        storage::with_storage_layer,
+        traits::{
+            fungible::Mutate,
+            tokens::{Preservation, Provenance},
+        },
+    };
+    use frame_system::pallet_prelude::*;
+    use sp_core::{H160, H256, U256};
+
+    use super::*;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    /// Configuration trait of this pallet.
+    #[pallet::config]
+    pub trait Config: frame_system::Config + pallet_ethereum::Config {
+        /// Overarching event type.
+        type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+        /// The EVM user account identifier type.
+        type EvmAccountId: Parameter
+            + Member
+            + MaybeSerializeDeserialize
+            + Debug
+            + MaybeDisplay
+            + Ord
+            + MaxEncodedLen
+            + Into<H160>;
+
+        /// Native currency.
+        type NativeCurrency: Inspect<Self::AccountId> + Mutate<Self::AccountId>;
+
+        /// EVM currency.
+        type EvmCurrency: Inspect<Self::EvmAccountId>;
+
+        /// The converter to determine how the balance amount should be converted from native
+        /// to evm currency.
+        type BalanceConverter: Convert<NativeBalanceOf<Self>, EvmBalanceOf<Self>>;
+
+        /// The pot native bridge account.
+        type PotNativeBridge: Get<Self::AccountId>;
+
+        /// The pot evm bridge account.
+        type PotEvmBridge: Get<Self::EvmAccountId>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+    }
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub(super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Balances were swapped.
+        BalancesSwapped {
+            /// The account id balances withdrawed from.
+            from: T::AccountId,
+            /// The withdrawed balances amount.
+            withdrawed_amount: NativeBalanceOf<T>,
+            /// The account id balances deposited to.
+            to: T::EvmAccountId,
+            /// The deposited balances amount.
+            deposited_amount: EvmBalanceOf<T>,
+            /// The corresponding transaction hash executed in evm.
+            evm_trx_hash: H256,
```

Comment:
```suggestion
            evm_transaction_hash: H256,
```

Reply:
Rationale: this `trx` abbreviation is competing with `tx`, and I'd rather avoid the hassle of picking one at all.

Reply:
Done - https://github.com/humanode-network/humanode/pull/1449/commits/c68cf72734fa580ce72a01d61d4a3a96b54c8639